Source/WTF/wtf/NativePromise.h

```diff
@@ -0,0 +1,1036 @@
+    using PromiseType = NativePromise;
+
+    [[nodiscard]] static Ref<Producer> create(const char* resolveSite)
+    {
+        return adoptRef(*new Producer(resolveSite));
+    }
+
+    template<typename ResolveValueType_, typename = std::enable_if<!std::is_void_v<ResolveValueT>>>
+    void resolve(ResolveValueType_&& resolveValue, const char* resolveSite)
+    {
+        static_assert(std::is_convertible_v<ResolveValueType_, ResolveValueT>, "resolve() argument must be implicitly convertible to NativePromise's ResolveValueT");
+        Locker lock { m_lock };
+        PROMISE_LOG("%s resolving NativePromise (%p created at %s)", resolveSite, this, m_creationSite);
+        ASSERT(isNothing());
+        if (!isNothing()) {
+            PROMISE_LOG("%s ignored already resolved or rejected NativePromise (%p created at %s)", resolveSite, this, m_creationSite);
+            return;
+        }
+        m_result = std::forward<ResolveValueType_>(resolveValue);
+        dispatchAll();
+    }
+
+    template<typename = std::enable_if<std::is_void_v<ResolveValueT>>>
+    void resolve(const char* resolveSite)
+    {
+        Locker lock { m_lock };
+        PROMISE_LOG("%s resolving NativePromise (%p created at %s)", resolveSite, this, m_creationSite);
+        ASSERT(isNothing());
+        if (!isNothing()) {
+            PROMISE_LOG("%s ignored already resolved or rejected NativePromise (%p created at %s)", resolveSite, this, m_creationSite);
+            return;
+        }
+        m_result = Result { };
+        dispatchAll();
+    }
+
+    template<typename RejectValueType_>
+    void reject(RejectValueType_&& rejectValue, const char* rejectSite)
+    {
+        static_assert(std::is_convertible_v<RejectValueType_, RejectValueT>, "reject() argument must be implicitly convertible to NativePromise's RejectValueT");
+        Locker lock { m_lock };
+        PROMISE_LOG("%s rejecting NativePromise (%p created at %s)", rejectSite, this, m_creationSite);
+        ASSERT(isNothing());
+        if (!isNothing()) {
+            PROMISE_LOG("%s ignored already resolved or rejected NativePromise (%p created at %s)", rejectSite, this, m_creationSite);
+            return;
+        }
+        m_result = Unexpected<RejectValueT>(std::forward<RejectValueType_>(rejectValue));
+        dispatchAll();
+    }
+
+    template<typename ResolveOrRejectValue_>
+    void resolveOrReject(ResolveOrRejectValue_&& result, const char* site)
+    {
+        Locker lock { m_lock };
+        ASSERT(isNothing());
+        if (!isNothing()) {
+            PROMISE_LOG("%s ignored already resolved or rejected NativePromise (%p created at %s)", site, this, m_creationSite);
+            return;
+        }
+        PROMISE_LOG("%s resolveOrRejecting NativePromise (%p created at %s)", site, this, m_creationSite);
+        m_result = std::forward<ResolveOrRejectValue_>(result);
+        dispatchAll();
+    }
+private:
+    explicit Producer(const char* creationSite)
```

Comment:
As discussed off-line, I reworked it so that Producer is now an independent class with no inheritance to NativePromise.
It can however, be converted to a consumer NativePromiser for ease